lib/semantics/resolve-labels.cc

```diff
@@ -0,0 +1,917 @@
+              "label '%" PRIu64 "' cannot be found"_err_en_US, label});
+    } else if (doTarget.parserCharBlock.begin() < position.begin()) {
+      // R1119
+      errorHandler.Say(position,
+          parser::MessageFormattedText{
+              "label '%" PRIu64 "' doesn't lexically follow DO stmt"_err_en_US,
+              label});
+    } else if (!InInclusiveScope(scopes, scope, doTarget.proxyForScope)) {
+      // C1133
+      errorHandler.Say(position,
+          parser::MessageFormattedText{
+              "label '%" PRIu64 "' is not in scope"_en_US, label});
+    } else if (!doTarget.labeledStmtClassificationSet.test(
+                   TargetStatementEnum::Do) &&
+        !doTarget.labeledStmtClassificationSet.test(
+            TargetStatementEnum::CompatibleDo)) {
+      errorHandler.Say(doTarget.parserCharBlock,
+          parser::MessageFormattedText{
+              "'%" PRIu64 "' invalid DO terminal statement"_err_en_US, label});
+    } else if (!doTarget.labeledStmtClassificationSet.test(
+                   TargetStatementEnum::Do)) {
+      errorHandler.Say(doTarget.parserCharBlock,
+          parser::MessageFormattedText{
+              "'%" PRIu64 "' invalid DO terminal statement"_en_US, label});
+    } else {
+      loopBodies.emplace_back(SkipLabel(position), doTarget.parserCharBlock);
+    }
+  }
+
+  CheckBranchesIntoDoBody(branches, labels, scopes, loopBodies, errorHandler);
+  CheckDoNesting(loopBodies, errorHandler);
+}
+
+// 6.2.5
+void CheckScopeConstraints(const SourceStmtList &stmts,
+    const TargetStmtMap &labels, const std::vector<ProxyForScope> &scopes,
+    parser::Messages &errorHandler) {
+  for (const auto stmt : stmts) {
+    const auto &label{stmt.parserLabel};
+    const auto &scope{stmt.proxyForScope};
+    const auto &position{stmt.parserCharBlock};
+    auto target{GetLabel(labels, label)};
+    if (!HasScope(target.proxyForScope)) {
+      errorHandler.Say(position,
+          parser::MessageFormattedText{
+              "label '%" PRIu64 "' was not found"_err_en_US, label});
+    } else if (!InInclusiveScope(scopes, scope, target.proxyForScope)) {
+      errorHandler.Say(position,
+          parser::MessageFormattedText{
+              "label '%" PRIu64 "' is not in scope"_en_US, label});
+    }
+  }
+}
+
+void CheckBranchTargetConstraints(const SourceStmtList &stmts,
+    const TargetStmtMap &labels, parser::Messages &errorHandler) {
+  for (const auto stmt : stmts) {
+    const auto &label{stmt.parserLabel};
+    auto branchTarget{GetLabel(labels, label)};
+    if (HasScope(branchTarget.proxyForScope)) {
+      if (!branchTarget.labeledStmtClassificationSet.test(
+              TargetStatementEnum::Branch) &&
+          !branchTarget.labeledStmtClassificationSet.test(
+              TargetStatementEnum::CompatibleBranch)) {
+        errorHandler.Say(branchTarget.parserCharBlock,
+            parser::MessageFormattedText{
+                "'%" PRIu64 "' not a branch target"_err_en_US, label});
```

Comment:
Can this error message be improved, maybe to something like "statement at label %d cannot be used as the target of a branch"?